github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17434168?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17434168?buildTab=artifacts#meta) on refs/heads/master @ [d2480d711e8e](https://github.com/cockroachdb/pebble/commits/d2480d711e8e):


```
        iter55 = snap39.NewIter("jqcxsmb", "kcyik@326", 2 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */) // <nil> #5184
        iter55.SeekLT("btqso@202", "") // [false] <nil> #5186
        iter55.SetBounds("lbejevajk@217", "rujztsnbh@28") // <nil> #5187
        iter55.SeekLT("rujztsnbh@28", "") // [true,"rqxbtqso","ffdkdzijy",<no range>] <nil> #5188
        iter55.Prev("") // [true,"rhbtqso","ffdkdzijy",<no range>] <nil> #5189
        iter55.Next("ynfnqj@226") // [valid,"rqxbtqso","ffdkdzijy",<no range>] <nil> #5191
        iter55.SetBounds("btqso@151", "btqsoatxbn@153") // <nil> #5192
        iter55.SeekPrefixGE("btqso@54") // [true,"btqso@54","ffdkdzijy",<no range>] <nil> #5193
        iter55.First() // [true,"btqso@151","ulqwpeeyczazqbkwaoe",<no range>] <nil> #5194
        iter55.First() // [true,"btqso@151","ulqwpeeyczazqbkwaoe",<no range>] <nil> #5195
        iter55.Next("") // [true,"btqso@149","ffdkdzijy",<no range>] <nil> #5196
        // 08:15:16.176 FATAL: levelIter L6: upper bound violation: &{bvgdithsdn@211#0,SET {xyebzzajdppcz %!s(*base.LazyFetcher=<nil>)}} > btqsoatxbn@153
        // 08:15:16.176 FATAL: goroutine 74 [running]:
        // 08:15:16.176 FATAL: runtime/debug.Stack()
        // 08:15:16.176 FATAL: 	GOROOT/src/runtime/debug/stack.go:24 +0x5e
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble.(*levelIter).verify(0xc000719df0, 0xc000976418)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:631 +0x293
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble.(*levelIter).SeekLT(0xc000719df0, {0xc000046e29, 0xe, 0x17}, 0x0)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:708 +0x145
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble/internal/invalidating.(*iter).SeekLT(0xc000b8ae60, {0xc000046e29?, 0x17?, 0xc000046e29?}, 0xe?)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/internal/invalidating/external/com_github_cockroachdb_pebble/internal/invalidating/iter.go:121 +0x28
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble.(*mergingIter).seekLT(0xc000718638, {0xc000046e29?, 0x48584a?, 0x15706c0?}, 0x10000c0000dbc20?, 0x0)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:1069 +0x2a8
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble.(*mergingIter).SeekLT(0xc000718638, {0xc000046e29?, 0x48?, 0x42?}, 0xe0?)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:1119 +0x51
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).SeekLT(0xc000734030, {0xc000046e29, 0xe, 0x17}, 0x0?)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/internal/keyspan/external/com_github_cockroachdb_pebble/internal/keyspan/interleaving_iter.go:371 +0xfa
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble.(*Iterator).iterLastWithinBounds(0xc000718008)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:2112 +0x48
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble.(*Iterator).Last(0xc000718008)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1687 +0xf0
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).Last.func1()
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:100 +0x25
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).withRetry(0xc00074bcb0, 0xc0000dbd78)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:47 +0x2b
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble/metamorphic.(*retryableIter).Last(0xc0000dbde8?)
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/retryable.go:100 +0x47
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble/metamorphic.(*iterLastOp).run(0xc0004377f0, 0xc00036fa60?, {0xc00031ad80?, 0xc0000b60c0?, 0x0?})
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/ops.go:1592 +0x5d
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble/metamorphic.(*Test).runOp(0xc00011e900?, 0xc0000dbee0?, {0xc00031ad80?, 0x0?, 0x0?})
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/test.go:435 +0x16f
        // 08:15:16.176 FATAL: github.com/cockroachdb/pebble/metamorphic.Execute.func1()
        // 08:15:16.176 FATAL: 	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:630 +0x116
        // 08:15:16.176 FATAL: golang.org/x/sync/errgroup.(*Group).Go.func1()
        // 08:15:16.176 FATAL: 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x56
        // 08:15:16.176 FATAL: created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 38
        // 08:15:16.176 FATAL: 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x96
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/241024-081512.033977643534/random-002 --try-to-reduce -v
        --- FAIL: TestMeta/execution/random-002 (0.90s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1729757712033650571 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: PEBBLE-285

With #4108:

```
panic: prefix changed bit for key "bxbtqso@275" (row 64) is false, expected true [prev key was "bvgdithsdn@211"]
        panic: Release called on a BufferPool with in-use buffers

goroutine 1781 [running]:
github.com/cockroachdb/pebble/sstable/block.(*BufferPool).Release(0x14000b35bc8)
        /Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/block/buffer_pool.go:172 +0xd4
panic({0x102905ac0?, 0x140009091d0?})
        /usr/local/go/src/runtime/panic.go:785 +0x124
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).enqueueDataBlock(0x140005ea008, {0x1400089e000, 0x3f2, 0x3f2}, {{0x14000403060?, 0x102503418?, 0x14000402c50?}, 0x43?}, {0x14000403070, 0xb, ...})
        /Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/colblk_writer.go:642 +0x2a8
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).flushDataBlockWithoutNextKey(0x140005ea008, {0x14000402c50, 0x7, 0x10})
        /Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/colblk_writer.go:607 +0x138
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).AddWithForceObsolete(0x140005ea008, {{0x14000402c50?, 0x7?, 0x10?}, 0xdcf?}, {0x1400012c318, 0x9, 0x18}, 0x0)
        /Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/colblk_writer.go:395 +0x2f0
github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable(0x140005b5710, {0x1029a0760, 0x140005ea008})
        /Users/jackson/go/src/github.com/cockroachdb/pebble/internal/compact/run.go:203 +0x4e0
github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable(0x140005b5710, {0x437, 0x2, {0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0}, ...}}, ...)
        /Users/jackson/go/src/github.com/cockroachdb/pebble/internal/compact/run.go:143 +0x178
github.com/cockroachdb/pebble.(*DB).compactAndWrite(0x140003a6008, 0x582, 0x14000b35b08, {0x0, 0x0, 0x0}, 0x7)
        /Users/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:2913 +0x4f4
github.com/cockroachdb/pebble.(*DB).runCompaction(0x140003a6008, 0x582, 0x14000b35b08)
        /Users/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:2808 +0x4e4
github.com/cockroachdb/pebble.(*DB).compact1(0x140003a6008, 0x14000b35b08, 0x0)
        /Users/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:2252 +0x11c
github.com/cockroachdb/pebble.(*DB).compact.func1({0x10299a550?, 0x140004291d0?})
        /Users/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:2166 +0xd4
runtime/pprof.Do({0x10299a4a0?, 0x102e06280?}, {{0x1400007f160?, 0x0?, 0x0?}}, 0x140004d7f78)
        /usr/local/go/src/runtime/pprof/runtime.go:51 +0x78
github.com/cockroachdb/pebble.(*DB).compact(0x0?, 0x1021bf3b4?, 0x1400097ac08?)
        /Users/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:2163 +0x60
created by github.com/cockroachdb/pebble.(*DB).tryScheduleAutoCompaction in goroutine 1778
        /Users/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:1845 +0x34c
```

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17449187?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17449187?buildTab=artifacts#meta) on refs/heads/master @ [f498510f91ef](https://github.com/cockroachdb/pebble/commits/f498510f91ef):


```
        // 04:29:35.249 INFO: [JOB 5744] compacting(default) L5 [023001] (1.0KB) Score=209.71 + L6 [012327] (1004B) Score=0.04; OverlappingRatio: Single 0.96, Multi 0.00
        // 04:29:35.249 INFO: [JOB 5740] sstable deleted 022330
        // 04:29:35.249 INFO: [JOB 5744] compacting: sstable created 023008
        // 04:29:35.252 INFO: [JOB 5743] compacted(default) L5 [022299] (1.1KB) Score=210.77 + L6 [012280] (1.0KB) Score=0.04 -> L6 [023007] (954B), in 0.6s (0.6s total), output rate 1.5KB/s
        // 04:29:35.972 INFO: [JOB 5741] sstable deleted 012309
        // 04:29:35.972 INFO: [JOB 5745] compacting(default) L5 [022336] (1.0KB) Score=208.72 + L6 [012318] (1012B) Score=0.04; OverlappingRatio: Single 0.97, Multi 0.00
        // 04:29:35.973 INFO: [JOB 5745] compacting: sstable created 023009
        // 04:29:35.975 INFO: [JOB 5744] compacted(default) L5 [023001] (1.0KB) Score=209.71 + L6 [012327] (1004B) Score=0.04 -> L6 [023008] (924B), in 0.7s (0.7s total), output rate 1.2KB/s
        // 04:29:36.703 INFO: [JOB 5746] compacting(default) L5 [022583] (1.0KB) Score=207.68 + L6 [012319] (1004B) Score=0.04; OverlappingRatio: Single 0.97, Multi 0.00
        // 04:29:36.703 INFO: [JOB 5741] sstable deleted 022331
        // 04:29:36.703 INFO: [JOB 5746] compacting: sstable created 023010
        // 04:29:36.705 INFO: [JOB 5745] compacted(default) L5 [022336] (1.0KB) Score=208.72 + L6 [012318] (1012B) Score=0.04 -> L6 [023009] (928B), in 0.7s (0.7s total), output rate 1.2KB/s
        // 04:29:37.416 INFO: [JOB 5742] sstable deleted 012277
        // 04:29:37.416 INFO: [JOB 5747] compacting(default) L5 [022593] (1.0KB) Score=206.67 + L6 [012322] (1004B) Score=0.04; OverlappingRatio: Single 0.97, Multi 0.00
        // 04:29:37.416 INFO: [JOB 5747] compacting: sstable created 023011
        // 04:29:37.418 INFO: [JOB 5746] compacted(default) L5 [022583] (1.0KB) Score=207.68 + L6 [012319] (1004B) Score=0.04 -> L6 [023010] (924B), in 0.7s (0.7s total), output rate 1.3KB/s
        // 04:29:38.116 INFO: [JOB 5748] compacting(default) L5 [022624] (1.0KB) Score=205.65 + L6 [012323] (996B) Score=0.04; OverlappingRatio: Single 0.97, Multi 0.00
        // 04:29:38.116 INFO: [JOB 5742] sstable deleted 021962
        // 04:29:38.116 INFO: [JOB 5748] compacting: sstable created 023012
        // 04:29:38.118 INFO: [JOB 5747] compacted(default) L5 [022593] (1.0KB) Score=206.67 + L6 [012322] (1004B) Score=0.04 -> L6 [023011] (924B), in 0.7s (0.7s total), output rate 1.3KB/s
        // 04:29:38.796 INFO: [JOB 5743] sstable deleted 012278
        // 04:29:38.796 INFO: [JOB 5749] compacting(default) L5 [022985] (1.0KB) Score=204.64 + L6 [012326] (1004B) Score=0.04; OverlappingRatio: Single 0.97, Multi 0.00
        // 04:29:38.796 INFO: [JOB 5749] compacting: sstable created 023013
        // 04:29:38.799 INFO: [JOB 5748] compacted(default) L5 [022624] (1.0KB) Score=205.65 + L6 [012323] (996B) Score=0.04 -> L6 [023012] (920B), in 0.7s (0.7s total), output rate 1.3KB/s
        // 04:29:39.478 INFO: [JOB 5750] compacting(default) L5 [022950] (1.0KB) Score=203.61 + L6 [012324] (1012B) Score=0.04; OverlappingRatio: Single 0.97, Multi 0.00
        // 04:29:39.478 INFO: [JOB 5743] sstable deleted 021963
        // 04:29:39.478 INFO: [JOB 5750] compacting: sstable created 023014
        // 04:29:39.480 INFO: [JOB 5749] compacted(default) L5 [022985] (1.0KB) Score=204.64 + L6 [012326] (1004B) Score=0.04 -> L6 [023013] (924B), in 0.7s (0.7s total), output rate 1.3KB/s
        // 04:29:40.169 INFO: [JOB 5744] sstable deleted 012280
        // 04:29:40.169 INFO: [JOB 5751] compacting(default) L5 [022966] (1.0KB) Score=202.58 + L6 [012325] (1020B) Score=0.04; OverlappingRatio: Single 0.97, Multi 0.00
        // 04:29:40.170 INFO: [JOB 5751] compacting: sstable created 023015
        // 04:29:40.172 INFO: [JOB 5750] compacted(default) L5 [022950] (1.0KB) Score=203.61 + L6 [012324] (1012B) Score=0.04 -> L6 [023014] (928B), in 0.7s (0.7s total), output rate 1.3KB/s
        // 04:29:40.869 INFO: [JOB 5752] compacting(default) L5 [020814] (1.0KB) Score=201.53 + L6 [012222] (1.0KB) Score=0.04; OverlappingRatio: Single 0.98, Multi 0.00
        // 04:29:40.869 INFO: [JOB 5744] sstable deleted 022299
        // 04:29:40.869 INFO: [JOB 5752] compacting: sstable created 023016
        // 04:29:40.871 INFO: [JOB 5751] compacted(default) L5 [022966] (1.0KB) Score=202.58 + L6 [012325] (1020B) Score=0.04 -> L6 [023015] (932B), in 0.7s (0.7s total), output rate 1.3KB/s
        // 04:29:41.565 INFO: [JOB 5745] sstable deleted 012327
        // 04:29:41.567 INFO: [JOB 5752] compacted(default) L5 [020814] (1.0KB) Score=201.53 + L6 [012222] (1.0KB) Score=0.04 -> L6 [023016] (974B), in 0.7s (0.7s total), output rate 1.4KB/s
        // 04:29:42.250 INFO: [JOB 5753] compacting(default) L5 [020819] (1.0KB) Score=200.46 + L6 [012224] (1.0KB) Score=0.04; OverlappingRatio: Single 0.98, Multi 0.00
        // 04:29:42.250 INFO: [JOB 5754] compacting(default) L5 [020821] (1.0KB) Score=199.41 + L6 [012225] (1.0KB) Score=0.04; OverlappingRatio: Single 0.98, Multi 0.00
        // 04:29:42.251 INFO: [JOB 5754] compacting: sstable created 023017
        // 04:29:42.251 INFO: [JOB 5753] compacting: sstable created 023018
        // 04:29:42.961 INFO: [JOB 5737] flush error: pebble: empty table
        // 04:29:42.961 INFO: [JOB 5745] sstable deleted 023001
        // 04:29:42.963 INFO: [JOB 5754] compacted(default) L5 [020821] (1.0KB) Score=199.41 + L6 [012225] (1.0KB) Score=0.04 -> L6 [023017] (966B), in 0.7s (0.7s total), output rate 1.3KB/s
        // 04:29:43.647 INFO: [JOB 5746] sstable deleted 012318
        // 04:29:43.650 INFO: [JOB 5753] compacted(default) L5 [020819] (1.0KB) Score=200.46 + L6 [012224] (1.0KB) Score=0.04 -> L6 [023018] (970B), in 0.7s (2.1s total), output rate 1.3KB/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/241025-042513.922479273964/random-016 --try-to-reduce -v
        --- FAIL: TestMeta/execution/random-016 (270.13s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1729830313922840856 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/sstable/colblk/bitmap.go
+++ b/sstable/colblk/bitmap.go
@@ -9,12 +9,12 @@ import (
 	"io"
 	"math"
 	"math/bits"
-	"slices"
 	"strings"
 	"unsafe"
 
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble/internal/binfmt"
+	"github.com/cockroachdb/pebble/internal/invariants"
 	"github.com/cockroachdb/pebble/internal/treeprinter"
 )
 
@@ -280,6 +280,8 @@ func (b *BitmapBuilder) Set(i int) {
 	}
 	w := i >> 6 // divide by 64
 	for len(b.words) <= w {
+		// We append zeros because if b.words has additional capacity, it has
+		// not been zeroed.
 		b.words = append(b.words, 0)
 	}
 	b.words[w] |= 1 << uint(i&63)
@@ -294,7 +296,16 @@ func (b *BitmapBuilder) isZero(rows int) bool {
 
 // Reset resets the bitmap to the empty state.
 func (b *BitmapBuilder) Reset() {
-	clear(b.words)
+	if invariants.Sometimes(50) {
+		// Sometimes trash the bitmap with all ones to catch bugs that assume
+		// b.words is zeroed.
+		for i := 0; i < len(b.words); i++ {
+			b.words[i] = ^uint64(0)
+		}
+	}
+
+	// NB: We don't zero the contents of b.words. When the BitmapBuilder reuses
+	// b.words, it must ensure it zeroes the contents as necessary.
 	b.words = b.words[:0]
 	b.minNonZeroRowCount = 0
 }
@@ -339,8 +350,10 @@ func (b *BitmapBuilder) Invert(nRows int) {
 	b.minNonZeroRowCount = 1
 	// If the tail of b is sparse, fill in zeroes before inverting.
 	nBitmapWords := (nRows + 63) >> 6
-	if len(b.words) < nBitmapWords {
-		b.words = slices.Grow(b.words, nBitmapWords-len(b.words))
+	for len(b.words) < nBitmapWords {
+		// We append zeros because if b.words has additional capacity, it has
+		// not been zeroed.
+		b.words = append(b.words, 0)
 	}
 	b.words = b.words[:nBitmapWords]
 	for i := range b.words {

--- a/sstable/colblk/bitmap.go
+++ b/sstable/colblk/bitmap.go
@@ -226,9 +226,11 @@ func (b Bitmap) SeekUnsetBitLE(i int) int {
 	return (wordIdx << 6) + 63 - bits.LeadingZeros64(^word)
 }
 
+// summaryTableBounds returns the indexes of the bitmap words containing the
+// summary table. The summary table's words lie within [startOffset, endOffset).
 func (b Bitmap) summaryTableBounds() (startOffset, endOffset int) {
 	startOffset = (b.bitCount + 63) >> 6
-	endOffset = startOffset + startOffset>>6
+	endOffset = startOffset + (startOffset+63)>>6
 	return startOffset, endOffset
 }
 

--- a/sstable/colblk/bitmap_test.go
+++ b/sstable/colblk/bitmap_test.go
@@ -147,17 +147,31 @@ func TestBitmapRandom(t *testing.T) {
 	seed := uint64(time.Now().UnixNano())
 	t.Logf("seed: %d", seed)
 	rng := rand.New(rand.NewPCG(0, seed))
-	size := rng.IntN(4096) + 1
 
-	testWithProbability := func(t *testing.T, p float64) {
-		var builder BitmapBuilder
-		v := make([]bool, size)
-		for i := 0; i < size; i++ {
+	var builder BitmapBuilder
+	testWithProbability := func(t *testing.T, p float64, size int, invert bool) {
+		defer builder.Reset()
+		rng := rand.New(rand.NewPCG(0, seed))
+
+		// Sometimes Set +1 bit to test the common pattern of excluding the last
+		// row from a data block.
+		buildSize := size + rng.IntN(5)
+		v := make([]bool, buildSize)
+		for i := range v {
 			v[i] = rng.Float64() < p
 			if v[i] {
 				builder.Set(i)
 			}
 		}
+
+		// Sometimes invert the bitmap.
+		if invert {
+			builder.Invert(size)
+			for i := 0; i < size; i++ {
+				v[i] = !v[i]
+			}
+		}
+
 		data := make([]byte, builder.Size(size, 0))
 		_ = builder.Finish(0, size, 0, data)
 		bitmap, endOffset := DecodeBitmap(data, 0, size)
@@ -219,16 +233,26 @@ func TestBitmapRandom(t *testing.T) {
 		}
 	}
 
+	fixedSizes := []int{1, 2, 3, 4, 16, 63, 64, 65, 128, 129, 256, 257, 1024, 1025, 4096}
 	fixedProbabilities := []float64{0.00001, 0.0001, 0.001, 0.1, 0.5, 0.9999}
 	for _, p := range fixedProbabilities {
 		t.Run(fmt.Sprintf("p=%05f", p), func(t *testing.T) {
-			testWithProbability(t, p)
+			for _, sz := range fixedSizes {
+				t.Run(fmt.Sprintf("size=%d", sz), func(t *testing.T) {
+					t.Run("invert", func(t *testing.T) {
+						testWithProbability(t, p, sz, true /* invert */)
+					})
+					t.Run("no-invert", func(t *testing.T) {
+						testWithProbability(t, p, sz, false /* invert */)
+					})
+				})
+			}
 		})
 	}
 	for i := 0; i < 10; i++ {
 		p := rng.ExpFloat64() * 0.1
 		t.Run(fmt.Sprintf("p=%05f", p), func(t *testing.T) {
-			testWithProbability(t, p)
+			testWithProbability(t, p, rng.IntN(4096)+1, rng.IntN(2) == 1)
 		})
 	}
 }

--- a/sstable/colblk/bitmap_test.go
+++ b/sstable/colblk/bitmap_test.go
@@ -233,7 +233,7 @@ func TestBitmapRandom(t *testing.T) {
 		}
 	}
 
-	fixedSizes := []int{1, 2, 3, 4, 16, 63, 64, 65, 128, 129, 256, 257, 1024, 1025, 4096}
+	fixedSizes := []int{1, 2, 3, 4, 16, 63, 64, 65, 128, 129, 256, 257, 1024, 1025, 4096, 4097, 8012, 8200}
 	fixedProbabilities := []float64{0.00001, 0.0001, 0.001, 0.1, 0.5, 0.9999}
 	for _, p := range fixedProbabilities {
 		t.Run(fmt.Sprintf("p=%05f", p), func(t *testing.T) {
@@ -252,7 +252,7 @@ func TestBitmapRandom(t *testing.T) {
 	for i := 0; i < 10; i++ {
 		p := rng.ExpFloat64() * 0.1
 		t.Run(fmt.Sprintf("p=%05f", p), func(t *testing.T) {
-			testWithProbability(t, p, rng.IntN(4096)+1, rng.IntN(2) == 1)
+			testWithProbability(t, p, rng.IntN(8200)+1, rng.IntN(2) == 1)
 		})
 	}
 }

